### PR TITLE
Add parsing and config for path implementation

### PR DESCRIPTION
**Pull request by bowdencm**
No changes to functionality (except for throwing an error on a "SET PATH" query), but building out the base for utilizing the path in function resolution.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    public String getName()
+    {
+        return "SET PATH";
+    }
+
+    @Override
+    public ListenableFuture<?> execute(SetPath statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine, List<Expression> parameters)
+    {
+        //TODO: implementation
+        throw new PrestoException(GENERIC_INTERNAL_ERROR, format("temporary exception to test set path: %s", statement.getPath()));
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java]: `throw new UnsupportedOperationException("SET PATH not implemented yet");`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CurrentPath.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+public class CurrentPath
+        extends Expression
+{
+    public CurrentPath()
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CurrentPath.java]: this seems unused.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.Optional;
+
+import static com.google.common.base.MoreObjects.toStringHelper;
+
+public class SetPath
+        extends Statement
+{
+    private final String path;
+
+    public SetPath(String path)
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java]: unused

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        //avoid splitting on commas contained within quotes as part of catalog/schema names
+        String[] schemas = rawPath.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");
+
+        for (String schema : schemas) {
+            String catalogName;
+            String schemaName;
+            SqlPathElement element;
+
+            //avoid splitting on periods contained within quotes as part of catalog/schema names
+            String[] individual = schema.split("\\.(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: Would be nice if we can rename `individual` to be more descriptive. Some options: `parts`, `catalogAndSchema`, etc.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+package com.facebook.presto.sql;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+
+public class SqlPath
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: `public final class`

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.Objects;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SqlPathElement
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java]: `public final class`

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            //avoid splitting on periods contained within quotes as part of catalog/schema names
+            String[] individual = schema.split("\\.(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");
+
+            if (individual.length > 1) {
+                catalogName = individual[0];
+                schemaName = individual[1];
+                element = new SqlPathElement(catalogName, schemaName);
+            }
+            else {
+                element = new SqlPathElement(schema);
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: We can use `new SqlPathElement(null, schema);` and we can get rid of this constructor.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    public int hashCode()
+    {
+        return Objects.hash(catalogName, schemaName);
+    }
+
+    @Override
+    public String toString()
+    {
+        if (catalogName != null) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java]: We can simply do `return format("%s.%s", catalogName, schemaName);` (format will handle null properly)

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 9]
```diff
@@ -102,6 +105,7 @@ public Session(
         this.source = requireNonNull(source, "source is null");
         this.catalog = requireNonNull(catalog, "catalog is null");
         this.schema = requireNonNull(schema, "schema is null");
+        this.path = path;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: Please add `requireNonNull` for `path`.

File: presto-main/src/test/java/com/facebook/presto/server/TestQuerySessionSupplier.java [hunk 10]
```diff
@@ -156,7 +163,27 @@ public void testInvalidTimeZone()
...
+        correctValues.add(new SqlPathElement("a\"\"terrible", "\"thing!@#$%^&*()\""));
+        List<SqlPathElement> correctList = correctValues.build();
+
+        SqlPath path = new SqlPath("normal.schema,"
+                + "\"who.uses.periods\".\"in.schema.names\","
+                + "\"same,deal\".\"with,commas\","
+                + "a\"\"terrible.\"thing!@#$%^&*()\"");
+        List<SqlPathElement> testList = path.getParsedPath();
+
+        assertEquals(correctList, testList);
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/server/TestQuerySessionSupplier.java]: `assertEquals` takes the `actual` result as the first parameter, but the expected result goes first here (the result is the same, but just to be consistent with the API of this method).

Let's rename `correctList` -> `expected` and `testList` -> `actual`, and we can do `assertEquals(actual, expected)`.

**Comment by nezihyigitbasi** [comment]
I made a quick pass and left some comments. Please also rebase to resolve the conflict.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java (hunk 8 above)
Comment (bowdencm) [reply presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java]: My concern with this was having it formatted to "null.<schema>" or something of the sort - is that not an issue?

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 11]
```diff
@@ -95,6 +95,7 @@ statement
     | EXECUTE identifier (USING expression (',' expression)*)?         #execute
     | DESCRIBE INPUT identifier                                        #describeInput
     | DESCRIBE OUTPUT identifier                                       #describeOutput
+    | SET PATH string                                                  #setPath
```
Comment (martint) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: Path has a well defined structure in the SQL grammar, so we should be parsing that here. For reference:

```
<path specification> ::= PATH <schema name list>
<schema name list> ::= <schema name> [ { <comma> <schema name> }... ]
<schema name> ::= [ <catalog name> <period> ] <unqualified schema name>
<unqualified schema name> ::= <identifier>
<catalog name> ::= <identifier>
```

Comment (haozhun) [reply presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: I have no context on this specific PR. But in general, we have opted to parse things out of the parser in order to have better error message.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java (hunk 8 above)
Comment (nezihyigitbasi) [reply presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java]: I am fine either way as `toString` will be used during debugging.

Comment (bowdencm) [reply presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java]: A similar implementation to this was necessary in the parser, so I think it makes sense to keep it like this for consistency's sake.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.sql;
+
+import io.airlift.configuration.Config;
+
+public class SqlEnvironmentConfig
+{
+    //TODO: determine an actual default path
+    private String path = " ";
```
Comment (elonazoulay) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java]: Would there be an issue having the default path as a space? Maybe better to keep it empty or null.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+    {
+        return Objects.hash(parsedPath);
+    }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("rawPath", rawPath)
+                .add("parsedPath", parsedPath != null ? Iterables.toString(parsedPath) : null)
```
Comment (elonazoulay) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: Nit: if the paths contain commas it might be clearer to re-escape the quotes, quote each path element and join with commas.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 14]
```diff
@@ -1386,6 +1388,13 @@ public void testShowGrants()
                 new ShowGrants(false, Optional.empty()));
     }
 
+    @Test
+    public void testSetPath()
+    {
+        assertStatement("SET PATH iLikeToEat.apples, andBananas",
+                new SetPath(ImmutableList.of(new PathElement("iLikeToEat", "apples"), new PathElement("andBananas"))));
+    }
```
Comment (elonazoulay) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: Add some invalid paths, like a path with 3 elements.

File: presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+    @Override
+    public String getName()
+    {
+        return "SET PATH";
+    }
+
+    @Override
+    public ListenableFuture<?> execute(SetPath statement, TransactionManager transactionManager, Metadata metadata, AccessControl accessControl, QueryStateMachine stateMachine, List<Expression> parameters)
+    {
+        //TODO: implementation
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/execution/SetPathTask.java]: unnecessary comment (exception message already says that)

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 16]
```diff
@@ -1134,6 +1136,15 @@ public Void visitShowGrants(ShowGrants node, Integer indent)
             return null;
         }
 
+        @Override
+        public Void visitSetPath(SetPath node, Integer indent)
+        {
+            builder.append("SET PATH ");
+            builder.append(node.getPath().stream().map(element -> format(element.toString())).collect(joining(", ")));
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: this is a bit complicated, so we can split it up to make it more readable:

```
String setPath = node.getPath().stream()
        .map(element -> format(element.toString()))
        .collect(joining(", "));
builder.append(setPath);
```

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 17]
```diff
@@ -1171,4 +1182,13 @@ private static void appendAliasColumns(StringBuilder builder, List<Identifier> c
                     .append(')');
         }
     }
+
+    private static void appendPathElement(StringBuilder builder, PathElement element)
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: this method is not used, please remove.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+
+    public SetPath(NodeLocation location, List<PathElement> path)
+    {
+        this(Optional.of(location), path);
+    }
+
+    public SetPath(Optional<NodeLocation> location, List<PathElement> path)
+    {
+        super(location);
+
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java]: you can remove the newline after the `super` call.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java (hunk 12 above)
Comment (bowdencm) [reply presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java]: It can't be null because it's passed to constructors that have requireNonNull(). Otherwise, it's not utilized - it'll be changed once we actually have code that uses the path for function resolution.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql;
+
+import io.airlift.configuration.Config;
+
+public class SqlEnvironmentConfig
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java]: We usually add tests for config classes like this one. For example, see how `TestNodeMemoryConfig` tests defaults and explicit configs for `NodeMemoryConfig`.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+        this.rawPath = path;
+    }
+
+    @JsonProperty
+    public String getRawPath()
+    {
+        return rawPath;
+    }
+
+    public List<SqlPathElement> getParsedPath()
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: What do you think about naming `parsedPath` as `path`, so this method becomes `getPath`, which I think will simplify the name a bit? We can also rename `rawPath` as `pathString`.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        return parsedPath;
+    }
+
+    private void parsePath()
+    {
+        //TODO: add further verification once default path is determined
+        ImmutableList.Builder<SqlPathElement> builder = ImmutableList.builder();
+        //avoid splitting on commas contained within quotes as part of catalog/schema names
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: As discussed offline, please add a reference comment to the SO answer.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+{
+    private final String catalogName;
+    private final String schemaName;
+
+    @JsonCreator
+    public SqlPathElement(
+            @JsonProperty("catalogName") String catalogName,
+            @JsonProperty("schemaName") String schemaName)
+    {
+        this.catalogName = catalogName;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPathElement.java]: Since `catalogName` can be null, we can mark it as nullable:

```
@JsonProperty("catalogName") @Nullable String catalogName
```

File: presto-main/src/test/java/com/facebook/presto/server/TestQuerySessionSupplier.java [hunk 23]
```diff
@@ -156,7 +163,27 @@ public void testInvalidTimeZone()
...
+        correctValues.add(new SqlPathElement("a\"\"terrible", "\"thing!@#$%^&*()\""));
+        List<SqlPathElement> expected = correctValues.build();
+
+        SqlPath path = new SqlPath("normal.schema,"
+                + "\"who.uses.periods\".\"in.schema.names\","
+                + "\"same,deal\".\"with,commas\","
+                + "a\"\"terrible.\"thing!@#$%^&*()\"");
+        List<SqlPathElement> actual = path.getParsedPath();
+
+        assertEquals(actual, expected);
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/server/TestQuerySessionSupplier.java]: nit: you can just do `assertEquals(path.getParsedPath(), expected);`

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/CurrentPath.java [hunk 24]
```diff
@@ -22,7 +22,6 @@
 public class CurrentPath
         extends Expression
 {
-
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/CurrentPath.java]: unnecessary whitespace change.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java (hunk 20 above)
Comment (bowdencm) [reply presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: I wanted to avoid this because it creates situations in which a line of code could look like: 
```session.getPath().getPath();```
The choice was to reserve the name ```getPath()``` for retrieving the SqlPath object, and that from there, the option was to get either the parsed version or the raw version.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java (hunk 19 above)
Comment (bowdencm) [reply presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java]: Would it make sense to add a test class now, or would that be better in a later commit once the default is determined?

File: presto-main/src/test/java/com/facebook/presto/sql/TestSqlEnvironmentConfig.java [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+
+import io.airlift.configuration.testing.ConfigAssertions;
+import org.testng.annotations.Test;
+
+public class TestSqlEnvironmentConfig
+{
+    @Test
+    public void testDefaults()
+    {
+        ConfigAssertions.assertRecordedDefaults(ConfigAssertions.recordDefaults(SqlEnvironmentConfig.class)
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/sql/TestSqlEnvironmentConfig.java]: static import the assert methods on this line, so this becomes:

```
assertRecordedDefaults(recordDefaults(SqlEnvironmentConfig.class).setPath(" "));
```

File: presto-main/src/test/java/com/facebook/presto/sql/TestSqlEnvironmentConfig.java [hunk 26]
```diff
@@ -0,0 +1,37 @@
...
+                .setPath(" "));
+    }
+
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/sql/TestSqlEnvironmentConfig.java]: we should add a unit test for valid configs:

```
@Test
public void testExplicitPropertyMappings()
{
    Map<String, String> properties = new ImmutableMap.Builder<String, String>()
            .put("sql.path", "a.b, c.d")
            .build();

    SqlEnvironmentConfig expected = new SqlEnvironmentConfig()
            .setPath("a.b, c.d");

    assertFullMapping(properties, expected);
}
```

File: presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+{
+    //TODO: determine an actual default path
+    private String path = " ";
+
+    @Config("path")
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java]: Please rename this config as `sql.path`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarCurrentPath.java [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+        extends ExpressionRewriteRuleSet
+{
+    public DesugarCurrentPath()
+    {
+        super(createRewrite());
+    }
+
+    private static ExpressionRewriter createRewrite()
+    {
+        return (expression, context) -> ExpressionTreeRewriter.rewriteWith(new com.facebook.presto.sql.tree.ExpressionRewriter<Void>()
```
Comment (elonazoulay) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DesugarCurrentPath.java]: No need to use the fully qualified name since you import this class already.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public final class PathElement
+        extends Node
+{
+    private final Optional<String> catalog;
+    private final String schema;
+
+    public PathElement(String schema)
```
Comment (elonazoulay) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java]: The current convention in presto is to minimize the amount of constructors, I see why this would be convenient but in general it's easier to reason about classes with a single constructor. Just use the Optional.empty() for missing arguments.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java [hunk 30]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        this(Optional.empty(), Optional.empty(), schema);
+    }
+
+    public PathElement(NodeLocation location, String schema)
+    {
```
Comment (elonazoulay) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java]: ditto

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+        this(Optional.of(location), Optional.empty(), schema);
+    }
+
+    public PathElement(String catalog, String schema)
```
Comment (elonazoulay) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java]: ditto

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java [hunk 32]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        this(Optional.empty(), Optional.of(catalog), schema);
+    }
+
+    public PathElement(NodeLocation location, String catalog, String schema)
```
Comment (elonazoulay) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java]: ditto

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java [hunk 33]
```diff
@@ -0,0 +1,89 @@
...
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.util.Objects.requireNonNull;
+
+public class SetPath
+        extends Statement
+{
+    private final List<PathElement> path;
+
+    public SetPath(List<PathElement> path)
```
Comment (elonazoulay) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java]: Same here - use a single constructor and set missing arguments to Optional.empty()

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java [hunk 34]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        this(Optional.empty(), path);
+    }
+
+    public SetPath(NodeLocation location, List<PathElement> path)
```
Comment (elonazoulay) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java]: ditto

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java [hunk 35]
```diff
@@ -25,27 +25,27 @@
 public class SetPath
         extends Statement
 {
-    private final List<PathElement> path;
+    private final PathSpecification pathSpecification;
 
-    public SetPath(List<PathElement> path)
+    public SetPath(PathSpecification pathSpecification)
```
Comment (elonazoulay) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java]: Same here: just use a single constructor and set the missing arguments to Optional.empty()

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java [hunk 36]
```diff
@@ -25,27 +25,27 @@
...
     {
-        this(Optional.empty(), path);
+        this(Optional.empty(), pathSpecification);
     }
 
-    public SetPath(NodeLocation location, List<PathElement> path)
+    public SetPath(NodeLocation location, PathSpecification pathSpecification)
```
Comment (elonazoulay) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java [hunk 37]
```diff
@@ -178,7 +178,12 @@ public Session createSession(QueryId queryId, SessionContext context, Optional<S
             sessionBuilder.setClientTransactionSupport();
         }
 
-        sessionBuilder.setPath(new SqlPath(path));
+        if (path.isPresent()) {
+            sessionBuilder.setPath(new SqlPath(path.get()));
+        }
+        else {
+            sessionBuilder.setPath(new SqlPath(null));
```
Comment (elonazoulay) [inline presto-main/src/main/java/com/facebook/presto/server/QuerySessionSupplier.java]: Should this be changed to use Optional instead of null? This way you can separate a null check from an empty path.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java [hunk 38]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.Optional;
+
+public class SqlEnvironmentConfig
+{
+    private Optional<String> path = Optional.empty();
+
+    @Config("sql.path")
+    public SqlEnvironmentConfig setPath(String path)
+    {
+        if (path == null) {
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlEnvironmentConfig.java]: this if/else is equivalent to `this.path = Optional.ofNullable(path);`

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 39]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.joining;
+
+public class SqlPath
+{
+    private List<SqlPathElement> parsedPath;
+    private Optional<String> rawPath;
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: mark `rawPath` as `final`.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 40]
```diff
@@ -0,0 +1,98 @@
...
+    public int hashCode()
+    {
+        return Objects.hash(parsedPath);
+    }
+
+    @Override
+    public String toString()
+    {
+        return getParsedPath().stream()
+                .map(element -> format(element.toString()))
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: this `format` call is redundant.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 41]
```diff
@@ -0,0 +1,98 @@
...
+
+        return parsedPath;
+    }
+
+    private void parsePath()
+    {
+        checkState(rawPath.isPresent(), "rawPath must be present to parse");
+
+        SqlParser parser = new SqlParser();
+        List<PathElement> parserPath = parser.createPathSpecification(rawPath.get()).getPath();
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: `parserPath` -> `pathSpecification`

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 42]
```diff
@@ -0,0 +1,98 @@
...
+
+        ImmutableList.Builder<SqlPathElement> builder = ImmutableList.builder();
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]:
```
this.parsedPath = pathSpecification.stream()
        .map(pathElement -> new SqlPathElement(pathElement.getCatalog(), pathElement.getSchema()))
        .collect(toImmutableList());
```

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java (hunk 40 above)
Comment (nezihyigitbasi) [reply presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]:
```
        return getParsedPath().stream()
                .map(SqlPathElement::toString)
                .collect(joining(", "));
```

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java [hunk 43]
```diff
@@ -1134,6 +1135,17 @@ public Void visitShowGrants(ShowGrants node, Integer indent)
             return null;
         }
 
+        @Override
+        public Void visitSetPath(SetPath node, Integer indent)
+        {
+            builder.append("SET PATH ");
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: `format` call is redundant, `map` call can be simplified to use method references:

```
builder.append(node.getPathSpecification().getPath().stream()
        .map(PathElement::toString)
        .collect(joining(", ")));
```

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java [hunk 44]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        this(Optional.empty(), Optional.of(catalog), schema);
+    }
+
+    public PathElement(NodeLocation location, Identifier catalog, Identifier schema)
+    {
+        this(Optional.of(location), Optional.of(catalog), schema);
+    }
+
+    public PathElement(Optional<NodeLocation> location, Optional<Identifier> catalog, Identifier schema)
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/PathElement.java]: can be private.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java [hunk 45]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        this(Optional.empty(), pathSpecification);
+    }
+
+    public SetPath(NodeLocation location, PathSpecification pathSpecification)
+    {
+        this(Optional.of(location), pathSpecification);
+    }
+
+    public SetPath(Optional<NodeLocation> location, PathSpecification pathSpecification)
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/SetPath.java]: can be private

File: presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java (hunk 43 above)
Comment (bowdencm) [reply presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java]: The changed code is one line beneath the code referenced in this comment.

File: presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java [hunk 46]
```diff
@@ -0,0 +1,91 @@
...
+import com.google.common.base.Joiner;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class SqlPath
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/sql/SqlPath.java]: We can also mark this class as `final`.

File: presto-main/src/test/java/com/facebook/presto/sql/TestSqlEnvironmentConfig.java [hunk 47]
```diff
@@ -0,0 +1,53 @@
...
+import static io.airlift.configuration.testing.ConfigAssertions.assertFullMapping;
+import static io.airlift.configuration.testing.ConfigAssertions.assertRecordedDefaults;
+
+public class TestSqlEnvironmentConfig
+{
+    @Test
+    public void testDefaults()
+    {
+        assertRecordedDefaults(ConfigAssertions.recordDefaults(SqlEnvironmentConfig.class)
```
Comment (nezihyigitbasi) [inline presto-main/src/test/java/com/facebook/presto/sql/TestSqlEnvironmentConfig.java]: static import `recordDefaults`.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/PathSpecification.java [hunk 48]
```diff
@@ -0,0 +1,82 @@
...
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class PathSpecification
```
Comment (nezihyigitbasi) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/PathSpecification.java]: We can mark this one as `final` too.

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/PathSpecification.java [hunk 49]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        this(Optional.empty(), path);
+    }
+
+    public PathSpecification(NodeLocation location, List<PathElement> path)
+    {
+        this(Optional.of(location), path);
+    }
+
+    public PathSpecification(Optional<NodeLocation> location, List<PathElement> path)
```
[the remark this answers is not in the document]
Comment (nezihyigitbasi) [reply presto-parser/src/main/java/com/facebook/presto/sql/tree/PathSpecification.java]: We should add `VisibleForTesting` to this constructor.

File: presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java [hunk 50]
```diff
@@ -1415,6 +1418,41 @@ public void testShowGrants()
...
+                            new PathElement(Optional.empty(), new Identifier("dummyValue"))))));
+            fail();
+        }
+        catch (RuntimeException e) {
+            //expected - schema can only be qualified by catalog
+        }
+
+        try {
+            SQL_PARSER.createStatement("SET PATH ", new ParsingOptions());
+        }
```
Comment (nezihyigitbasi) [inline presto-parser/src/test/java/com/facebook/presto/sql/parser/TestSqlParser.java]: We need a `fail();` right after `createStatement`.
